### PR TITLE
Execute task hint #3: validate hash

### PR DESCRIPTION
**Pull request by odesenfans**
Implemented a module that computes the hash of a Cairo program in Rust.

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
  - [ ] CHANGELOG has been updated.

### CODE REVIEW

**Comment by odesenfans** [comment]
The majority of the PR can be reviewed but the hint is not finished as we don't know exactly where to store programs inside the `Task` struct. I added a fake implementation for now to make the PR compile.

File: vm/src/hint_processor/builtin_hint_processor/bootloader/program_hash.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    maybe_relocatable: &MaybeRelocatable,
+) -> Result<FieldElement, ProgramHashError> {
+    let felt = maybe_relocatable
+        .get_int_ref()
+        .ok_or(ProgramHashError::InvalidProgramData)?;
+    felt_to_field_element(felt)
+}
+
+#[allow(dead_code)] // TODO: remove
+/// Computes the hash of a program.
```
Comment (notlesh) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/program_hash.rs:96]:
```suggestion
/// Computes the Pedersen hash of a program.
```

Or some other hint about what the hash impl is

File: vm/src/hint_processor/builtin_hint_processor/bootloader/program_hash.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+///     assert len(data) >= 1, f"len(data) for hash chain computation must be >= 1; got: {len(data)}."
+///     return functools.reduce(lambda x, y: hash_func(y, x), data[::-1])
+fn compute_hash_chain<'a, I>(
+    data: I,
+    hash_func: HashFunction,
+) -> Result<FieldElement, HashChainError>
+where
+    I: Iterator<Item = &'a FieldElement> + DoubleEndedIterator,
+{
+    match data.copied().rev().reduce(|x, y| hash_func(&y, &x)) {
```
Comment (notlesh) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/program_hash.rs:53]: Looks good, does `rev()` come from `data[::-1]`, or something else?

File: vm/src/hint_processor/builtin_hint_processor/bootloader/program_hash.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    let builtin_name = builtin
+        .name()
+        .strip_suffix("_builtin")
+        .unwrap_or(builtin.name());
+
+    FieldElement::from_byte_slice_be(builtin_name.as_bytes())
+        .map_err(|_| ProgramHashError::InvalidProgramBuiltin(builtin.name()))
+}
+
+/// The `value: FieldElement` is `pub(crate)` and there is no accessor.
```
Comment (notlesh) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/program_hash.rs:74]: Is there an upstream change that makes sense to fix this?

**Review by notlesh — COMMENTED** [review]
Looks good so far :+1:

**Review by notlesh — APPROVED** [review]

**Pull request merged**